header-translator/src/main.rs

```diff
@@ -0,0 +1,311 @@
+            version_min,
+            "-isysroot",
+            sdk.path.to_str().unwrap(),
+        ])
+        .parse()
+        .unwrap();
+
+    println!("status: initialized translation unit {:?}", sdk.platform);
+
+    dbg!(&tu);
+    dbg!(tu.get_target());
+    dbg!(tu.get_memory_usage());
+    dbg!(tu.get_diagnostics());
+
+    // let dbg_file = |file: File<'_>| {
+    //     dbg!(
+    //         &file,
+    //         file.get_module(),
+    //         file.get_skipped_ranges(),
+    //         file.is_include_guarded(),
+    //         // file.get_includes(),
+    //         // file.get_references(),
+    //     );
+    // };
+    //
+    // dbg_file(tu.get_file(&header).unwrap());
+    // dbg_file(tu.get_file(&dir.join("NSAccessibility.h")).unwrap());
+    // let cursor_file = tu.get_file(&dir.join("NSCursor.h")).unwrap();
+    // dbg_file(cursor_file);
+
+    let entity = tu.get_entity();
+
+    dbg!(&entity);
+    dbg!(entity.get_availability());
+
+    let framework_dir = sdk.path.join("System/Library/Frameworks");
+    entity.visit_children(|entity, _parent| {
+        if let Some(location) = entity.get_location() {
+            if let Some(file) = location.get_file_location().file {
+                let path = file.get_path();
+                if let Ok(path) = path.strip_prefix(&framework_dir) {
+                    let mut components = path.components();
+                    let library = components
+                        .next()
+                        .expect("components next")
+                        .as_os_str()
+                        .to_str()
+                        .expect("component to_str")
+                        .strip_suffix(".framework")
+                        .expect("framework fileending");
+
+                    let path = components.as_path();
+                    let name = path
+                        .file_stem()
+                        .expect("path file stem")
+                        .to_string_lossy()
+                        .to_owned();
+
+                    f(&library, &name, entity);
+                }
+            }
+        }
+        EntityVisitResult::Continue
+    });
+}
+
+fn output_files(
+    output_path: &Path,
+    files: impl IntoIterator<Item = (String, RustFile)>,
+    format_incrementally: bool,
+) -> fmt::Result {
+    let declared: Vec<_> = files
+        .into_iter()
+        .map(|(name, file)| {
+            let (declared_types, tokens) = file.finish();
+
+            let mut path = output_path.join(&name);
+            path.set_extension("rs");
+
+            let output = if format_incrementally {
+                run_rustfmt(tokens)
+            } else {
+                tokens.into()
+            };
+
+            fs::write(&path, output).unwrap();
+
+            (name, declared_types)
+        })
+        .collect();
+
+    let mut tokens = String::new();
+
+    for (name, _) in &declared {
+        writeln!(tokens, "pub(crate) mod {name};")?;
```

Comment:
Have you thought about using [`quote`](https://crates.io/crates/quote)? It's been my experience that it helps when actually generating the rust to verify correctness.

Reply:
Yeah, I started using that, but found that it was difficult to generate "good looking" Rust code (`rustfmt` can only go so far when things are nested in macros), and it wasn't really satisfactory wrt. performance.